system-contracts/test/MsgValueSimulator.spec.ts

```diff
@@ -0,0 +1,194 @@
+    const value = 1;
+    const balanceBefore = await l2EthToken.balanceOf(msgSender);
+    const contractBalanceBefore = await l2EthToken.balanceOf(mockContract.address);
+
+    await expect(
+      extraAbiCaller.fallback({
+        data: encodeExtraAbiCallerCalldata(
+          messageValueSimulator.address,
+          BigNumber.from(0),
+          [value.toString(), mockContract.address, "0"],
+          "0x"
+        ),
+      })
+    )
+      .to.emit(mockContract, "Called")
+      .withArgs(value, "0x");
+
+    const balanceAfter = await l2EthToken.balanceOf(msgSender);
+    expect(balanceBefore.sub(value)).to.equal(balanceAfter);
+    const contractBalanceAfter = await l2EthToken.balanceOf(mockContract.address);
+    expect(contractBalanceBefore.add(value)).to.equal(contractBalanceAfter);
+  });
+
+  it("send 2^127 wei", async () => {
+    const value = BigNumber.from(2).pow(127);
+    const balanceBefore = await l2EthToken.balanceOf(msgSender);
+    const contractBalanceBefore = await l2EthToken.balanceOf(mockContract.address);
+
+    await expect(
+      extraAbiCaller.fallback({
+        data: encodeExtraAbiCallerCalldata(
+          messageValueSimulator.address,
+          BigNumber.from(0),
+          [value.toString(), mockContract.address, "0"],
+          "0x"
+        ),
+      })
+    )
+      .to.emit(mockContract, "Called")
+      .withArgs(value, "0x");
+
+    const balanceAfter = await l2EthToken.balanceOf(msgSender);
+    expect(balanceBefore.sub(value)).to.equal(balanceAfter);
+    const contractBalanceAfter = await l2EthToken.balanceOf(mockContract.address);
+    expect(contractBalanceBefore.add(value)).to.equal(contractBalanceAfter);
+  });
+
+  it("revert with reentry", async () => {
+    const balanceBefore = await l2EthToken.balanceOf(msgSender);
+    const contractBalanceBefore = await l2EthToken.balanceOf(mockContract.address);
+
+    await expect(
+      extraAbiCaller.fallback({
+        data: encodeExtraAbiCallerCalldata(
+          messageValueSimulator.address,
+          BigNumber.from(0),
+          ["0x1", messageValueSimulator.address, "0"],
+          "0x"
+        ),
+      })
+    ).to.be.reverted;
+
+    const balanceAfter = await l2EthToken.balanceOf(msgSender);
+    expect(balanceBefore).to.equal(balanceAfter);
+
+    const contractBalanceAfter = await l2EthToken.balanceOf(mockContract.address);
+    expect(contractBalanceBefore).to.equal(contractBalanceAfter);
+  });
+
+  it("revert more than balance", async () => {
+    const balanceBefore = await l2EthToken.balanceOf(msgSender);
+    const contractBalanceBefore = await l2EthToken.balanceOf(mockContract.address);
+    const value = balanceBefore.add(1);
+
+    await expect(
+      extraAbiCaller.fallback({
+        data: encodeExtraAbiCallerCalldata(
+          messageValueSimulator.address,
+          BigNumber.from(0),
+          [value.toString(), mockContract.address, "0"],
+          "0x"
+        ),
+      })
+    ).to.be.reverted;
```

Comment:
Also check that the revert is empty, please

Reply:
Not sure how to do that, could you point me in the right direction?

Reply:
Doesn't something like `revertedWithoutReason` work?